File: src/main/java/com/guardanis/blank/tools/PendingEvents.java
package com.guardanis.blank.tools;

import java.util.HashMap;

public class PendingEvents {

    private static PendingEvents instance;
    public static PendingEvents getInstance() {
        if(instance == null)
            instance = new PendingEvents();

        return instance;
    }

    private HashMap<String, Runnable> pendingEvents = new HashMap<String, Runnable>();

    protected PendingEvents() { }

    public void register(String key, Runnable event) {
        pendingEvents.put(key, event);
    }

    public void unregister(String key) {
        pendingEvents.put(key, null);
    }

    public void trigger(String key) {
        Runnable event = pendingEvents.get(key);

        if(event != null){
            event.run();
            unregister(key);
        }
    }

    public boolean has(String key) {
        return pendingEvents.get(key) != null;
    }

}

File: src/main/java/com/guardanis/blank/tools/PendingEvents.kt
package com.guardanis.blank.tools

import java.util.HashMap

class PendingEvents protected constructor() {

    private val pendingEvents = HashMap<String, () -> Unit>()

    fun register(key: String, event: () -> Unit) {
        pendingEvents.put(key, event)
    }

    fun unregister(key: String) {
        pendingEvents.remove(key)
    }

    fun trigger(key: String) {
        pendingEvents[key]
                ?.let({ it.invoke() })
    }

    fun has(key: String): Boolean {
        return pendingEvents[key] != null
    }

    companion object {

        private var sharedInstance: PendingEvents? = null
        fun getInstance(): PendingEvents {
            if (sharedInstance == null)
                sharedInstance = PendingEvents()

            return sharedInstance!!
        }
    }
}